### DISCUSSION TITLE
Sending same embedding file (rag) multiple times increases token usage

### DISCUSSION THREAD

**Discussion by bgeneto** [open-webui/open-webui#5586] [General]
# Bug Report

## Installation Method

Docker

## Environment

- **Open WebUI Version:** v0.3.23
- **Ollama (if applicable):** n/a

- **Operating System:** Debian 12 6.1
- **Browser (if applicable):** Edge v129

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior

The system should only send or stream each file or collection (RAG embedding) once in a chat. To achieve this, a function should be implemented to analyze and preprocess the `files` and retain only one instance of each unique `collection_name`. This is crucial as sending duplicate `collection_name` entries significantly increases the input token count, resulting in substantial cost implications.

## Actual Behavior

The current system sends the same file or collection (RAG embedding) multiple times, effectively doubling (or even tripling) the actual number of input tokens.

## Description

**Bug Summary**

A bug has been identified where selecting a document or collection using the '#' character in a chat, followed by one or more consecutive questions about the context, results in the same document or collection being listed or sent multiple times in the input body. This is demonstrated in the example below, where the same filename appears twice.

```
{'stream': True, 'model': 'gpt-4o-mini', 'stream_options': {'include_usage': True}, 'messages': [{'role': 'user', 'content': 'what is pipelines according to context only?'}], 'temperature': 0.15, 'files': [{'type': 'file', 'collection_name': 'c0aabf58356294a210c649fbb22c599f50530928a02955fbd082fb000f6ccff', 'name': 'pipelinestutorial', 'title': 'open-webui-pipelines-tutorial', 'filename': 'open-webui-pipelines-tutorial.txt', 'content': {'tags': [{'name': 'open-webui'}]}, 'user_id': '01503731-0a49-4033-bdc0-8c1bc3b75fd6', 'timestamp': 1726908132, 'status': 'processed'}, {'type': 'file', 'collection_name': 'c0aabf58356294a210c649fbb22c599f50530928a02955fbd082fb000f6ccff', 'name': 'pipelinestutorial', 'title': 'open-webui-pipelines-tutorial', 'filename': 'open-webui-pipelines-tutorial.txt', 'content': {'tags': [{'name': 'open-webui'}]}, 'user_id': '01503731-0a49-4033-bdc0-8c1bc3b75fd6', 'timestamp': 1726908132, 'status': 'processed'}], 'session_id': 'uLGPka0fPoIniChP000b', 'chat_id': '4f93c8bc-e088-480e-bf3d-77ef929e9b3a', 'id': 'cae24965-ea47-4c96-bcd0-72b34984b733'}
```

## Reproduction Details

**Steps to Reproduce:**

Just select a document/collection using the '#' char in a chat and then ask one or more consecutive questions about the context. Inspect the inlet/body contents to confirm duplication. 

## Logs and Screenshots

**Browser Console Logs:**
None

**Docker Container Logs:**
None

**Screenshots/Screen Recordings (if applicable):**
None

## Additional Information
None

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by tjbck** [comment]
Unsure what you meant here, could you elaborate a bit more?

**Reply by bgeneto** [reply]
I've noticed a higher than expected input token usage while including a document with #. 

Then I decide to create a pipeline just to inspect what was sent and, to my surprise, the included document was present twice in the body. 

In fact, any time we use a collection or document and interact (several times) in the chat, the same document seems to be included multiple times, probability resulting in additional costs for input tokens (context).

This can be confirmed by inspecting the `body` and `type` = `file`.

What do you think about this duplicated body contents?

**Reply by mykola-mmm** [reply]
@bgeneto @tjbck I experience the same issue with tag v0.5.4

I have created a collection with a single .txt file.
when I include this collection into the chat and send a request I get such response:

```json
{
    "id": "a2f5f6f7-11d7-45ba-944d-980404760d9b",
    "role": "assistant",
    "content": "I'm not sure what specific information you're looking for regarding \"ads.\" Could you please provide more details or clarify your question?",
    "timestamp": 1736693198,
    "sources": [
        {
            "source": {
                "id": "747f97e5-1ee8-409f-8c1e-91e2c57dd9ce",
                "meta": {
                    "name": "test_document_upload",
                    "content_type": "application/octet-stream",
                    "size": 18,
                    "collection_name": "d380f587-be48-4be3-9873-78411563b0a2"
                },
                "created_at": 1736692920,
                "updated_at": 1736692920,
                "collection": {
                    "name": "test",
                    "description": "test"
                },
                "name": "test_document_upload",
                "description": "test - test",
                "type": "file",
                "status": "processed"
            },
            "document": [
                "test AI openweb ui",
                "test AI openweb ui",
                "test AI openweb ui"
            ],
            "metadata": [
                {
                    "created_by": "e3fd63c4-f279-455d-87d1-195ec2338f7b",
                    "embedding_config": "{\"engine\": \"\", \"model\": \"sentence-transformers/all-MiniLM-L6-v2\"}",
                    "file_id": "747f97e5-1ee8-409f-8c1e-91e2c57dd9ce",
                    "hash": "a1c37461cb048b0ef0876acdb5e3e216d834b4d24b8866d551e4cf63c935ada7",
                    "name": "test_document_upload",
                    "source": "test_document_upload",
                    "start_index": 0
                },
                {
                    "created_by": "e3fd63c4-f279-455d-87d1-195ec2338f7b",
                    "embedding_config": "{\"engine\": \"\", \"model\": \"sentence-transformers/all-MiniLM-L6-v2\"}",
                    "file_id": "747f97e5-1ee8-409f-8c1e-91e2c57dd9ce",
                    "hash": "a1c37461cb048b0ef0876acdb5e3e216d834b4d24b8866d551e4cf63c935ada7",
                    "name": "test_document_upload",
                    "source": "test_document_upload",
                    "start_index": 0
                },
                {
                    "created_by": "e3fd63c4-f279-455d-87d1-195ec2338f7b",
                    "embedding_config": "{\"engine\": \"\", \"model\": \"sentence-transformers/all-MiniLM-L6-v2\"}",
                    "file_id": "747f97e5-1ee8-409f-8c1e-91e2c57dd9ce",
                    "hash": "a1c37461cb048b0ef0876acdb5e3e216d834b4d24b8866d551e4cf63c935ada7",
                    "name": "test_document_upload",
                    "source": "test_document_upload",
                    "start_index": 0
                }
            ],
            "distances": [
                0.7292678071425029,
                0.9268915064874929,
                0.9308538351242461
            ]
        }
    ]
}
```

What is also strange here is that all 3 documents have different distances (I expect them to be the same because the documents are the same)
